### DISCUSSION TITLE
`Promise<ReactNode>` return type annotation not working with react 19 types

### DISCUSSION THREAD

**Discussion by stefanprobst** [vercel/next.js#67365] [Help] [answered]
### Summary

i am still unclear how to correctly type the return type of async components in react 19.

the following snippet (see the [example](https://github.com/stefanprobst/issue-react-types/blob/main/index.tsx)) errors:

```tsx
import type { ReactNode } from "react";

function Page() {
  // 'PageContent' cannot be used as a JSX component.
  // Its type '() => Promise<ReactNode>' is not a valid JSX element type.
  //   Type '() => Promise<ReactNode>' is not assignable to type '(props: any) => ReactNode'.
  //     Type 'Promise<ReactNode>' is not assignable to type 'ReactNode'.
  //       Type 'Promise<ReactNode>' is not assignable to type 'Promise<AwaitedReactNode>'.
  //         Type 'ReactNode' is not assignable to type 'AwaitedReactNode'.
  //           Type 'Promise<AwaitedReactNode>' is not assignable to type 'AwaitedReactNode'.ts(2786)
  return <PageContent />
}

async function PageContent(): Promise<ReactNode> {
  return <div>Check</div>
}
```

---

however, when using `Promise<Awaited<ReactNode>>` instead of `Promise<ReactNode>` it seems to work, but i have not seen that documented anywhere. is that the recommended approach? (maybe @eps1lon has more insight here? thanks!)

### Example

https://github.com/stefanprobst/issue-react-types

**Comment by eps1lon** [comment, marked as answer]
> when using Promise<Awaited<ReactNode>> instead of Promise<ReactNode> it seems to work

This is currently the only workaround. Either https://github.com/microsoft/TypeScript/issues/59111 gets addressed or we fix it at the `@types/react` level. It's a bit more involved to fix it before React 19 is stable so I'm not going to pursue it until then. Others are free to contribute. I added tests as a starting point: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69970

**Reply by stefanprobst** [reply]
thank you!

**Reply by Amraa9656** [reply]
hi i use const FooterLayout = async ({ locale }: { locale: string }): Promise<Awaited<ReactNode>> but this is not working ?

**Comment by lacymorrow** [comment]
Encountering this issue with next@15

**Reply by Netail** [reply]
Add `/// <reference types="react/experimental" />` to your app's global types

RE; https://github.com/facebook/react/issues/28718#issuecomment-2033952643

**Reply by booi-gg** [reply]
Did you find a solution?

**Reply by Netail** [reply]
Also, seems like @types/react v19 seems to have resolved the issue for us

**Reply by eps1lon** [reply]
It should work if you omit the return type but explicitly annotating `Promise<React.ReactNode>` as the return type should still fail until https://github.com/DefinitelyTyped/DefinitelyTyped/pull/71379 lands.

**Reply by manucian-official** [reply]
gg
